src/python/pants/java/nailgun_executor.py

```diff
@@ -153,7 +153,7 @@ def run(this, stdout=None, stderr=None, stdin=None, cwd=None):
 
   def _check_nailgun_state(self, new_fingerprint):
     running = self.is_alive()
-    updated = self.needs_restart(new_fingerprint) or self.cmd != self._distribution.java
```

Comment:
Why is this changing? I don't think it should be?

Reply:
I could delve more into why the restarting is occurring, because to start at least `self.cmd` is a list, not a string like `self._distribution.java`.

Reply:
Would be good to see whether this is still necessary.

Reply:
I think `cmd` is a string representation of `command` (https://github.com/pantsbuild/pants/blob/e61a164938b1f7da3dff19c3ae44f0a2b579d33f/src/python/pants/java/executor.py#L60), but that doesn't clear up why we're comparing that string to `self._distribution.java`, which is the location of java being used.

I think `self.cmd != self._distribution.java` would only be False if the cmd was to just run java without any arguments.

Reply:
It is still necessary, if you run e.g. `./pants test tests/python/pants_test/backend/jvm/tasks/jvm_compile/rsc: -- -vs -k test_executing_multi_target_binary_nonhermetic` it will reliably fail with the error described in the OP without this change, and that's not a fluke. Since the git log of that line only shows moving it from elsewhere (the first commit from `git log -L 156,156:src/python/pants/java/nailgun_executor.py` is `0d6207481c0e0a952258b741cdc4f2200bc3529c`), I don't know how to evaluate this. It would be great if there were comments or tests about it, but in the worst case if we start seeing random pantsd issues and #6579 hasn't been merged yet (which I suspect would make all of this much more clear), we know what to fix.

Reply:
(I could see that being used as a testing hack to invalidate dummy `java` processes, for example, but I don't know where to find more context about this)

Reply:
I was wrong, looking at the wrong level of nesting. `self.cmd` in this context is from the ProcessManager, not the Runner. It's the first element in the command line.

https://github.com/pantsbuild/pants/blob/70977ef064305b78406a627e07f4dae3a60e4ae4/src/python/pants/pantsd/process_manager.py#L293-L298

Reply:
I find it confusing that cmd has so many meanings in that file.

Reply:
Yep!!!!